jobmon_server/src/jobmon/server/web/routes/v2/fsm/array.py

```diff
@@ -349,3 +349,106 @@ async def get_array_max_concurrently_running(
             status_code=StatusCodes.OK,
         )
     return resp
+
+
+@api_v1_router.post("/array/{array_id}/transition_to_killed")
+@api_v2_router.post("/array/{array_id}/transition_to_killed")
+async def transition_to_killed(array_id: int, request: Request) -> Any:
+    """Transition TIs from KILL_SELF  to ERROR_FATAL.
+
+    Also mark parent Taskswith status=ERROR_FATAL if they're in a killable state.
+    """
+    structlog.contextvars.bind_contextvars(array_id=array_id)
+
+    data = cast(Dict, await request.json())
+    batch_num = data["batch_number"]
+
+    # 1) Acquire locks on the parent Tasks, and set them to ERROR_FATAL
+    #    if they're in a killable state.
+    #    This is analogous to how transition_to_launched locks tasks
+    #    that are INSTANTIATING and sets them to LAUNCHED.
+
+    with SessionLocal() as session:
+        with session.begin():
+            # Find Task IDs belonging to TIs in this array & batch
+            task_ids_query = (
+                select(TaskInstance.task_id)
+                .where(
+                    TaskInstance.array_id == array_id,
+                    TaskInstance.array_batch_num == batch_num,
+                    TaskInstance.status == TaskInstanceStatus.KILL_SELF,
+                )
+                .execution_options(synchronize_session=False)
+            )
+            task_ids = session.execute(task_ids_query).scalars().all()
+
+            # We'll define "killable" Task states. Adjust as appropriate.
+            killable_task_states = (
```

Comment:
Would we want any of the other Task statuses included in the killable states? e.g. instantiated, registered etc.